### PR TITLE
integrated camera_factory in video_input

### PR DESCRIPTION
**Pull request by Aleksa-M**

### CODE REVIEW

**Review by maxlou05 — COMMENTED** [review]
Reviewed

File: main_2024.py [hunk 1]
```diff
@@ -81,14 +84,24 @@ def main() -> int:
...
+        VIDEO_INPUT_WIDTH = config["video_input"]["width"]
+        VIDEO_INPUT_HEIGHT = config["video_input"]["height"]
+        if VIDEO_INPUT_OPTION == camera_factory.CameraOption.OPENCV:
+            VIDEO_INPUT_CAMERA_CONFIG = camera_opencv.ConfigOpenCV(
+                **config["video_input"]["camera_config"]
+            )
+        elif VIDEO_INPUT_OPTION == camera_factory.CameraOption.PICAM2:
+            VIDEO_INPUT_CAMERA_CONFIG = camera_picamera2.ConfigPiCamera2(
+                **config["video_input"]["camera_config"]
+            )
```
Comment (Xierumeng) [inline main_2024.py:98]: Might be worth putting these options into separate config files. For example:
`config.yaml`
```yaml
camera_enum: 0
# Other common camera options
camera_configs:
- camera_opencv.yaml
- camera_picam2.yaml
```
`main.py`
```py
camera_option = config["camera_enum"]
camera_config_path = config["camera_configs"][camera_option.value]
...

# Camera specific configuration
... = read_yaml.open_config(camera_config_path)
```

File: modules/video_input/video_input.py [hunk 2]
```diff
@@ -3,23 +3,72 @@
...
 
-    def run(self) -> "tuple[bool, image_and_time.ImageAndTime | None]":
+    @classmethod
+    def create(
+        cls,
+        camera_option: camera_factory.CameraOption,
+        width: int,
+        height: int,
+        config: camera_opencv.ConfigOpenCV | camera_picamera2.ConfigPiCamera2,
+        save_prefix: str,
```
Comment (Xierumeng) [inline modules/video_input/video_input.py:27]: Rename to `maybe_image_name` with type `str | None` . Then images are only logged if this variable is not none, and an empty string is valid input for logging.

File: modules/video_input/video_input.py [hunk 3]
```diff
@@ -3,23 +3,72 @@
...
         """
-        result, image = self.device.get_image()
+        result, image = self.__device.run()
         if not result:
+            self.__logger.warning("Failed to take image")
             return False, None
 
+        # If __save_prefix is not empty string, then save image
+        if self.__save_prefix:
+            self.__logger.save_image(image, self.__save_prefix)
```
Comment (Xierumeng) [inline modules/video_input/video_input.py:72]:
```py
if self.__maybe_image_name is not None:
    ...
```

File: modules/video_input/video_input_worker.py [hunk 4]
```diff
@@ -2,30 +2,56 @@
...
+from ..common.modules.camera import camera_opencv
+from ..common.modules.camera import camera_picamera2
+from ..common.modules.logger import logger
 
 
 def video_input_worker(
-    camera_name: "int | str",
-    period: float,
-    save_name: str,
+    period: int,
```
Comment (Xierumeng) [inline modules/video_input/video_input_worker.py:19]: Move period with the worker parameters. Also should be float.

File: modules/video_input/video_input_worker.py [hunk 5]
```diff
@@ -2,30 +2,56 @@
...
     output_queue: queue_proxy_wrapper.QueueProxyWrapper,
     controller: worker_controller.WorkerController,
 ) -> None:
     """
     Worker process.
 
-    camera_name is initial setting.
-    period is minimum period between loops.
-    save_name is path for logging.
+    period is the minimum period between image captures.
```
Comment (Xierumeng) [inline modules/video_input/video_input_worker.py:31]: Add seconds.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: main_2024.py (hunk 1 above)
Comment (maxlou05) [reply main_2024.py:98]: I feel like it is easier to keep it in one config file. It's easier to find, and these options aren't that big either.

File: modules/video_input/video_input_worker.py (hunk 4 above)
Comment (maxlou05) [reply modules/video_input/video_input_worker.py:19]: It currently is a worker parameter? I changed it to float

Comment (Xierumeng) [reply modules/video_input/video_input_worker.py]: I meant immediately above `output_queue` . Then the parameters for the class object and the worker are grouped separately.

File: config.yaml [hunk 6]
```diff
@@ -3,25 +3,37 @@
 queue_max_size: 10
 
 video_input:
-    camera_name: 0
     worker_period: 1.0  # seconds
-    save_prefix: "log_image"
+    camera_enum: 0  # enum values can be found in camera_factory.py
```
Comment (Xierumeng) [inline config.yaml:7]: Capitalize.

File: config.yaml [hunk 7]
```diff
@@ -3,25 +3,37 @@
...
+    width: 1920
+    height: 1200
+    # For camera_enum=0, use the CV camera config. For camera_enum=1, use the Picamera config
```
Comment (Xierumeng) [inline config.yaml:10]: OpenCV

Picamera2

File: config.yaml [hunk 8]
```diff
@@ -3,25 +3,37 @@
...
+    camera_config:  # For CV camera (regular cameras, enum 0)
```
Comment (Xierumeng) [inline config.yaml:11]: Move this comment into a line above for better separation.

OpenCV

File: config.yaml [hunk 9]
```diff
@@ -3,25 +3,37 @@
...
+        device_index: 0
+    # camera_config:  # For Picamera (PiCamera NoIR, enum 1)
```
Comment (Xierumeng) [inline config.yaml:13]: Move this comment into a line above for better separation.

Picamera2

File: config.yaml [hunk 10]
```diff
@@ -3,25 +3,37 @@
...
+    #     exposure_time: 250  # microseconds
+    #     analogue_gain: 64.0  # Sets ISO, 1.0 for normal, 64.0 for max, 0.0 for min
+    #     contrast: 1.0  # Contrast, 1.0 for nomral, 32.0 for max, 0.0 for min
+    #     lens_position: null  # Focal length, 1/m (0 for infinity, null for auto focus)
+    save_prefix: "log_image"  # Leave as empty string to not log any images
```
Comment (Xierumeng) [inline config.yaml:18]: Update this comment. Might be worth adding a boolean option for enabling/disabling logging.

File: main_2024.py [hunk 11]
```diff
@@ -81,14 +84,25 @@ def main() -> int:
...
         QUEUE_MAX_SIZE = config["queue_max_size"]
 
-        VIDEO_INPUT_CAMERA_NAME = config["video_input"]["camera_name"]
         VIDEO_INPUT_WORKER_PERIOD = config["video_input"]["worker_period"]
-        VIDEO_INPUT_SAVE_NAME_PREFIX = config["video_input"]["save_prefix"]
-        VIDEO_INPUT_SAVE_PREFIX = str(pathlib.Path(logging_path, VIDEO_INPUT_SAVE_NAME_PREFIX))
+        VIDEO_INPUT_OPTION = camera_factory.CameraOption(config["video_input"]["camera_enum"])
+        VIDEO_INPUT_WIDTH = config["video_input"]["width"]
+        VIDEO_INPUT_HEIGHT = config["video_input"]["height"]
+        match VIDEO_INPUT_OPTION:
```
Comment (Xierumeng) [inline main_2024.py:91]: Add default which does error handling and returns.

File: modules/video_input/video_input.py [hunk 12]
```diff
@@ -3,23 +3,72 @@
...
+        config: camera_opencv.ConfigOpenCV | camera_picamera2.ConfigPiCamera2,
+        maybe_image_name: str | None,
+        local_logger: logger.Logger,
+    ) -> "tuple[True, VideoInput] | tuple[False, None]":
+        """
+        camera_option specifies which camera driver to use.
+        width is the width of the images the camera takes in pixels.
+        height is the height of the images the camera takes in pixels.
+        camera_config specifies camera settings.
+        save_prefix is name of iamge log files. Leave as empty string for not logging the images.
```
Comment (Xierumeng) [inline modules/video_input/video_input.py:35]: Update this.

File: modules/video_input/video_input.py [hunk 13]
```diff
@@ -3,23 +3,72 @@
...
+        """
+        result, camera = camera_factory.create_camera(camera_option, width, height, config)
+        if not result:
+            local_logger.error(
+                f"camera factory failed. Current configs were: {camera_option=}, {width=}, {height=}, {config=}. Please try a different set of configs."
```
Comment (Xierumeng) [inline modules/video_input/video_input.py:40]: Capitalize. Remove: `Please try ...` , logging should be informational only, no suggestions.

File: modules/video_input/video_input.py [hunk 14]
```diff
@@ -3,23 +3,72 @@
...
+            )
+            return False, None
+        return True, VideoInput(cls.__create_key, camera, maybe_image_name, local_logger)
```
Comment (Xierumeng) [inline modules/video_input/video_input.py:43]: Add empty line above this line.

File: modules/video_input/video_input.py [hunk 15]
```diff
@@ -3,23 +3,72 @@
...
         """
         Returns a possible ImageAndTime with current timestamp.
         """
-        result, image = self.device.get_image()
+        result, image = self.__device.run()
         if not result:
+            self.__logger.warning("Failed to take image")
             return False, None
 
+        # If __maybe_image_name is not None, then save image
```
Comment (Xierumeng) [inline modules/video_input/video_input.py:70]: Remove comment.

File: tests/integration/test_video_input_hardware.py [hunk 16]
```diff
@@ -2,21 +2,38 @@
...
 from modules.video_input import video_input
 
 
-CAMERA = 0
+# Modify as needed
+CAMERA = camera_factory.CameraOption.OPENCV
+WIDTH = 1920
+HEIGHT = 1200
+CONFIG = camera_opencv.ConfigOpenCV(0)
+IAMGE_NAME = None  # Not saving any pictures
```
Comment (Xierumeng) [inline tests/integration/test_video_input_hardware.py:18]: Spelling.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: config.yaml [hunk 17]
```diff
@@ -3,25 +3,40 @@
...
+    camera_config:
+        device_index: 0
+    # PiCamera2 camera config (PiCamera NoIR, enum 1)
+    # camera_config:
+    #     exposure_time: 250  # microseconds
+    #     analogue_gain: 64.0  # Sets ISO, 1.0 for normal, 64.0 for max, 0.0 for min
+    #     contrast: 1.0  # Contrast, 1.0 for nomral, 32.0 for max, 0.0 for min
+    #     lens_position: null  # Focal length, 1/m (0 for infinity, null for auto focus)
+    log_images: true  # Set to true to save images
+    save_prefix: "log_image"  # Image name prefix when saving images
```
Comment (Xierumeng) [inline config.yaml:21]: This is not a prefix, see `Logger.save_image()` . Maybe: `image_name`

File: main_2024.py [hunk 18]
```diff
@@ -81,19 +84,38 @@ def main() -> int:
...
+            case camera_factory.CameraOption.OPENCV:
+                VIDEO_INPUT_CAMERA_CONFIG = camera_opencv.ConfigOpenCV(
+                    **config["video_input"]["camera_config"]
+                )
+            case camera_factory.CameraOption.PICAM2:
+                VIDEO_INPUT_CAMERA_CONFIG = camera_picamera2.ConfigPiCamera2(
+                    **config["video_input"]["camera_config"]
+                )
+            case _:
+                raise ValueError(f"Inputted an invalid camera option: {VIDEO_INPUT_OPTION}")
```
Comment (Xierumeng) [inline main_2024.py:101]: Log error and return instead.

File: main_2024.py [hunk 19]
```diff
@@ -81,19 +84,38 @@ def main() -> int:
...
+        if config["video_input"]["log_images"]:
```
Comment (Xierumeng) [inline main_2024.py:102]: Add empty line above this line.

File: main_2024.py [hunk 20]
```diff
@@ -81,19 +84,38 @@ def main() -> int:
...
+            VIDEO_INPUT_SAVE_PREFIX = None
+        else:
+            VIDEO_INPUT_SAVE_PREFIX = str(
+                pathlib.Path(logging_path, config["video_input"]["save_prefix"])
```
Comment (Xierumeng) [inline main_2024.py:106]: No, just use it directly, see logger.

File: main_2024.py [hunk 21]
```diff
@@ -81,19 +84,38 @@ def main() -> int:
...
+            )
```
Comment (Xierumeng) [inline main_2024.py:107]: I suggest ternary operation:
```py
VIDEO_INPUT_SAVE_PREFIX = if not config["video_input"]["log_images"] None else ...
```

File: modules/video_input/video_input.py [hunk 22]
```diff
@@ -3,23 +3,72 @@
...
+        config: camera_opencv.ConfigOpenCV | camera_picamera2.ConfigPiCamera2,
+        maybe_image_name: str | None,
+        local_logger: logger.Logger,
+    ) -> "tuple[True, VideoInput] | tuple[False, None]":
+        """
+        camera_option specifies which camera driver to use.
+        width is the width of the images the camera takes in pixels.
+        height is the height of the images the camera takes in pixels.
+        camera_config specifies camera settings.
+        maybe_image_name is name of iamge log files. Set to None to not log any images.
```
Comment (Xierumeng) [inline modules/video_input/video_input.py:35]: Update this to align with logger.

Spelling.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: tests/integration/test_video_input_hardware.py [hunk 23]
```diff
@@ -2,21 +2,38 @@
...
+from modules.common.modules.camera import camera_opencv
+from modules.common.modules.logger import logger
 from modules.video_input import video_input
 
 
-CAMERA = 0
+# Modify as needed
+CAMERA = camera_factory.CameraOption.OPENCV
+WIDTH = 1920
+HEIGHT = 1200
```
Comment (Xierumeng) [inline tests/integration/test_video_input_hardware.py:16]: Integration test is failing for me, I suggest a resolution of 640x480 .

File: tests/integration/test_video_input_worker.py [hunk 24]
```diff
@@ -6,14 +6,21 @@
...
 from modules.video_input import video_input_worker
 from modules import image_and_time
 from utilities.workers import queue_proxy_wrapper
 from utilities.workers import worker_controller
 
 
+# Modify these settings as needed
 VIDEO_INPUT_WORKER_PERIOD = 1.0
-CAMERA = 0
+CAMERA = camera_factory.CameraOption.OPENCV
```
Comment (Xierumeng) [inline tests/integration/test_video_input_worker.py:19]: Also failing here, but for some reason there is a conversion which results in this being `float` in the worker function.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

**Review by Xierumeng — APPROVED** [review]
Approved, thanks!

**Pull request merged**